File: goblinframework-embedded/src/main/java/org/goblinframework/embedded/core/provider/JdkHttpServletRequest.java
package org.goblinframework.embedded.core.provider;

import com.sun.net.httpserver.HttpExchange;
import org.goblinframework.core.util.StringUtils;
import org.goblinframework.embedded.core.servlet.AbstractHttpServletRequest;
import org.goblinframework.http.util.HttpUtils;
import org.jetbrains.annotations.NotNull;
import org.springframework.util.LinkedMultiValueMap;

import javax.servlet.ServletInputStream;
import java.net.InetSocketAddress;
import java.util.*;

public class JdkHttpServletRequest extends AbstractHttpServletRequest {

  private final HttpExchange exchange;
  private final JdkServletInputStream inStream;
  private final LinkedMultiValueMap<String, String> parameters;

  JdkHttpServletRequest(@NotNull HttpExchange exchange) {
    this.exchange = exchange;
    this.inStream = new JdkServletInputStream(this.exchange);
    this.parameters = new LinkedMultiValueMap<>();
    String qs = getQueryString();
    if (StringUtils.isNotEmpty(qs)) {
      HttpUtils.parseMultiQueryString(qs).forEach((name, values) -> {
        for (String value : values) {
          this.parameters.add(name, value);
        }
      });
    }
    String contentType = StringUtils.defaultString(getContentType());
    if ("POST".equalsIgnoreCase(getMethod())
        && contentType.contains("application/x-www-form-urlencoded")) {
      HttpUtils.parseMultiQueryString(inStream.requestBodyAsString()).forEach((name, values) -> {
        for (String value : values) {
          this.parameters.add(name, value);
        }
      });
    }
  }

  @Override
  public boolean isSecure() {
    return false;
  }

  @Override
  public ServletInputStream getInputStream() {
    return inStream;
  }

  @Override
  public String getParameter(String name) {
    return parameters.getFirst(name);
  }

  @Override
  public Enumeration<String> getParameterNames() {
    return Collections.enumeration(parameters.keySet());
  }

  @Override
  public String[] getParameterValues(String name) {
    List<String> values = parameters.get(name);
    if (values == null) {
      return null;
    }
    return values.toArray(new String[0]);
  }

  @Override
  public Map<String, String[]> getParameterMap() {
    Map<String, String[]> map = new HashMap<>();
    parameters.forEach((name, value) -> {
      String[] s = value.toArray(new String[0]);
      map.put(name, s);
    });
    return map;
  }

  @Override
  public String getScheme() {
    return exchange.getProtocol();
  }

  @Override
  public String getRemoteHost() {
    InetSocketAddress remote = exchange.getRemoteAddress();
    return remote == null ? null : remote.getHostName();
  }

  @Override
  public int getRemotePort() {
    InetSocketAddress remote = exchange.getRemoteAddress();
    return remote == null ? -1 : remote.getPort();
  }

  @Override
  public String getHeader(String name) {
    return exchange.getRequestHeaders().getFirst(name);
  }

  @Override
  public Enumeration<String> getHeaders(String name) {
    return Collections.enumeration(exchange.getRequestHeaders().get(name));
  }

  @Override
  public Enumeration<String> getHeaderNames() {
    return Collections.enumeration(exchange.getRequestHeaders().keySet());
  }

  @Override
  public String getMethod() {
    return exchange.getRequestMethod();
  }

  @Override
  public String getContextPath() {
    throw new UnsupportedOperationException();
  }

  @Override
  public String getQueryString() {
    return exchange.getRequestURI().getQuery();
  }

  @Override
  public String getRequestURI() {
    throw new UnsupportedOperationException();
  }
}

File: goblinframework-embedded/src/main/java/org/goblinframework/embedded/core/provider/JdkHttpServletRequest.kt
package org.goblinframework.embedded.core.provider

import com.sun.net.httpserver.HttpExchange
import org.apache.commons.io.IOUtils
import org.goblinframework.embedded.core.servlet.AbstractHttpServletRequest
import org.goblinframework.http.util.HttpUtils
import org.springframework.util.LinkedMultiValueMap
import java.util.*
import javax.servlet.ServletInputStream

class JdkHttpServletRequest(private val exchange: HttpExchange,
                            private val path: String,
                            private val query: String?) : AbstractHttpServletRequest() {

  companion object {
    private const val FORM_CONTENT_TYPE = "application/x-www-form-urlencoded"
  }

  private val inputStream = JdkServletInputStream(exchange)
  private val parameters = LinkedMultiValueMap<String, String>()

  init {
    query?.run {
      HttpUtils.parseMultiQueryString(this).forEach { (t, u) ->
        parameters.addAll(t, u.toList())
      }
    }
    if ("POST".equals(method, true)
        && (contentType ?: "").contains(FORM_CONTENT_TYPE, true)) {
      val content = inputStream.content()
      val s = IOUtils.toString(content, Charsets.UTF_8.name())
      HttpUtils.parseMultiQueryString(s).forEach { (t, u) ->
        parameters.addAll(t, u.toList())
      }
    }
  }

  override fun getMethod(): String {
    return exchange.requestMethod
  }

  override fun getInputStream(): ServletInputStream {
    return inputStream
  }

  override fun getRequestURI(): String {
    return path
  }

  override fun getQueryString(): String? {
    return query
  }

  override fun getHeaderNames(): Enumeration<String> {
    return Collections.enumeration(exchange.requestHeaders.keys)
  }

  override fun getHeaders(name: String): Enumeration<String> {
    val values = exchange.requestHeaders[name] ?: Collections.emptyList()
    return Collections.enumeration(values)
  }

  override fun getHeader(name: String): String? {
    return exchange.requestHeaders.getFirst(name)
  }

  override fun getParameterNames(): Enumeration<String> {
    return Collections.enumeration(parameters.keys)
  }

  override fun getParameter(name: String): String? {
    return parameters.getFirst(name)
  }

  override fun getParameterMap(): Map<String, Array<String>> {
    val map = mutableMapOf<String, Array<String>>()
    parameters.forEach { (t, u) -> map[t] = u.toTypedArray() }
    return Collections.unmodifiableMap(map)
  }

  override fun getParameterValues(name: String): Array<String>? {
    return parameters[name]?.toTypedArray()
  }
}